File: scripts/rig/debug-mist-eye-cycle.py
#!/usr/bin/env python3
"""Compare eye patch paints vs canonical token #1 render at eye coords."""

from __future__ import annotations

import json
import sys
from pathlib import Path

import numpy as np
from PIL import Image

REPO = Path(__file__).resolve().parents[2]
SCRIPTS = REPO / "scripts"
if str(SCRIPTS) not in sys.path:
    sys.path.insert(0, str(SCRIPTS))

from anim._canonical_token_source import canonical_grid_for_token, load_mint_records  # noqa: E402

PATCHES = REPO / "src" / "pages" / "AwakenDemo" / "mist-eye-patches.json"
OUT = REPO / "out" / "rig"

GRID = 64


def main() -> None:
    patches = json.loads(PATCHES.read_text(encoding="utf-8"))
    mint_records = load_mint_records()
    base = canonical_grid_for_token(1, mint_records)
    rgb = base[:, :, :3].astype(np.uint8)
    img = Image.fromarray(rgb, mode="RGB")
    OUT.mkdir(parents=True, exist_ok=True)
    img.save(OUT / "debug-token1-canonical-64.png")

    coords = [(p["x"], p["y"]) for p in patches["baselinePixels"] if p["x"] != 29]
    baseline_map = {(p["x"], p["y"]): p["rgba"][:3] for p in patches["baselinePixels"]}

    print("=== CHECK 4: canonical render vs patch baseline at eye coords ===")
    for x, y in coords:
        live = tuple(int(v) for v in rgb[y, x])
        patch = tuple(baseline_map.get((x, y), ()))
        match = live == patch
        print(f"  ({x},{y}) canonical={live} patch_baseline={patch} match={match}")

    print("\n=== CHECK 2/5: forced variant pixel changes on canonical ===")
    for name in patches["variants"]:
        out = rgb.copy()
        changed = []
        for p in patches["variants"][name]:
            x, y = p["x"], p["y"]
            before = tuple(int(v) for v in out[y, x])
            after = tuple(p["rgba"][:3])
            out[y, x] = after
            if before != after:
                changed.append({"coord": [x, y], "before": before, "after": after})
        Image.fromarray(out, mode="RGB").save(OUT / f"debug-token1-forced-{name}.png")
        print(f"  {name}: {len(changed)} changed pixels", changed[:4])

    # Simulate timer logic (15-30s intervals)
    print("\n=== CHECK 1: production timer params (15-30s, 90s wall) ===")
    import random

    random.seed(42)
    state = {"active": None, "next_ms": 0}
    events = []

    def schedule():
        return 15000 + random.random() * 15000

    state["next_ms"] = schedule()
    for ms in range(0, 90001, round(1000 / 12)):
        frame = ms * 12 // 1000
        if not state["active"] and ms >= state["next_ms"]:
            state["active"] = random.choice(list(patches["variants"].keys()))
            state["hold_until"] = frame + 24
            events.append({"t": ms, "frame": frame, "event": "start", "variant": state["active"]})
        elif state["active"] and frame >= state["hold_until"]:
            events.append({"t": ms, "frame": frame, "event": "end", "variant": state["active"]})
            state["active"] = None
            state["next_ms"] = ms + schedule()
    print(f"  swaps in 90s: {len([e for e in events if e['event']=='start'])}")
    for e in events:
        print(f"  {e}")


if __name__ == "__main__":
    main()

File: scripts/debug-mist-eye-cycle.mjs
/**
 * Diagnostic: mist eye-cycle patches vs live/canonical pixels at eye coords.
 * Usage: node scripts/debug-mist-eye-cycle.mjs
 */
import fs from "fs";
import { createCanvas, loadImage } from "@napi-rs/canvas";
import { unpackPixelIndicesFromHex } from "../src/lib/pixel-canvas.js";

const patches = JSON.parse(
  fs.readFileSync("src/pages/AwakenDemo/mist-eye-patches.json", "utf8"),
);
const mint = JSON.parse(fs.readFileSync("public/data/mint-data.json", "utf8"));
const record = mint.find((r) => r.tokenId === 1);

function rgbaAt(data, width, x, y) {
  const i = (y * width + x) * 4;
  return [data[i], data[i + 1], data[i + 2], data[i + 3]];
}

function renderCanonicalPng64() {
  // Use exported mockup if present, else decode from mint via simple approach:
  // Read reports PNG
  const mockup = "reports/robinhood/mockup/token1_native.png";
  if (fs.existsSync(mockup)) return mockup;
  return null;
}

function diffPixels(before, after, coords) {
  let changed = 0;
  const samples = [];
  for (const [x, y] of coords) {
    const b = rgbaAt(before, 64, x, y);
    const a = rgbaAt(after, 64, x, y);
    if (b.join() !== a.join()) {
      changed++;
      samples.push({ x, y, before: b, after: a });
    }
  }
  return { changed, samples };
}

async function samplePng(path) {
  const img = await loadImage(path);
  const c = createCanvas(64, 64);
  const ctx = c.getContext("2d");
  ctx.imageSmoothingEnabled = false;
  ctx.drawImage(img, 0, 0, 64, 64);
  return ctx.getImageData(0, 0, 64, 64).data;
}

function paintVariant(baseData, variantName) {
  const out = new Uint8ClampedArray(baseData);
  const pixels = patches.variants[variantName];
  for (const { x, y, rgba } of pixels) {
    const i = (y * 64 + x) * 4;
    out[i] = rgba[0];
    out[i + 1] = rgba[1];
    out[i + 2] = rgba[2];
    out[i + 3] = rgba[3];
  }
  return out;
}

function simulateTimer(seconds = 90) {
  const events = [];
  const state = {
    activeVariant: null,
    holdUntilFrame: 0,
    nextCycleAtMs: 5000, // start sooner for sim
  };
  const minIntervalMs = 5000;
  const maxIntervalMs = 10000;
  const holdFrames = 24;
  const fps = 12;

  function schedule() {
    return minIntervalMs + Math.random() * (maxIntervalMs - minIntervalMs);
  }
  state.nextCycleAtMs = schedule();

  const pool = Object.keys(patches.variants);
  for (let ms = 0; ms <= seconds * 1000; ms += 1000 / fps) {
    const frame = Math.floor(ms / (1000 / fps));
    const prev = state.activeVariant;
    if (!state.activeVariant && ms >= state.nextCycleAtMs) {
      state.activeVariant = pool[Math.floor(Math.random() * pool.length)];
      state.holdUntilFrame = frame + holdFrames;
      events.push({ type: "start", ms, frame, variant: state.activeVariant });
    } else if (state.activeVariant && frame >= state.holdUntilFrame) {
      events.push({ type: "end", ms, frame, variant: state.activeVariant });
      state.activeVariant = null;
      state.nextCycleAtMs = ms + schedule();
    }
    if (prev !== state.activeVariant && state.activeVariant && prev) {
      /* noop */
    }
  }
  return events;
}

console.log("=== CHECK 2: patch pixel distinctness vs baseline ===");
const baseline = patches.baselinePixels;
for (const [name, px] of Object.entries(patches.variants)) {
  let diffs = 0;
  const examples = [];
  for (const p of px) {
    const b = baseline.find((q) => q.x === p.x && q.y === p.y);
    const baseRgba = b?.rgba ?? null;
    if (!baseRgba || baseRgba.join() !== p.rgba.join()) {
      diffs++;
      if (examples.length < 3) examples.push({ coord: [p.x, p.y], base: baseRgba, variant: p.rgba });
    }
  }
  console.log(`${name}: ${diffs}/${px.length} pixels differ from baseline`, examples);
}

console.log("\n=== CHECK 1: timer simulation (5-10s intervals, 90s) ===");
const events = simulateTimer(90);
console.log(`events: ${events.length}`);
for (const e of events) console.log(e);

console.log("\n=== CHECK 4/5: canonical PNG path ===");
const pngPath = renderCanonicalPng64();
if (!pngPath) {
  console.log("No reports/robinhood/mockup/token1_native.png — checking baselinePixels only");
  console.log("baseline at eye coords:", baseline);
} else {
  const baseData = await samplePng(pngPath);
  const coords = patches.maskCoords;
  console.log("Live/canonical PNG pixels at eye coords:");
  for (const [x, y] of coords) {
    const live = rgbaAt(baseData, 64, x, y);
    const patch = baseline.find((p) => p.x === x && p.y === y)?.rgba;
    console.log(`  (${x},${y}) live=${JSON.stringify(live)} patch-baseline=${JSON.stringify(patch)}`);
  }

  console.log("\n=== CHECK 3/5: forced Chubby_CrossEyed pixel diff on canonical PNG ===");
  const forced = paintVariant(baseData, "Chubby_CrossEyed");
  const { changed, samples } = diffPixels(baseData, forced, coords.filter(([x]) => x !== 29));
  console.log(`changed pixels: ${changed}/8`, samples);

  const forcedStoned = paintVariant(baseData, "Chubby_Stoned");
  const stonedDiff = diffPixels(baseData, forcedStoned, coords.filter(([x]) => x !== 29));
  console.log("Chubby_Stoned changed:", stonedDiff.changed, stonedDiff.samples.slice(0, 4));
}

// mint-data grid at eye coords (role indices)
const grid = unpackPixelIndicesFromHex(record.pixelsHex);
console.log("\n=== mint-data role indices at eye coords ===");
for (const [x, y] of patches.maskCoords) {
  console.log(`  (${x},${y}) role=${grid[y * 64 + x]}`);
}
